ml-agents-envs/mlagents_envs/timers.py

```diff
@@ -232,9 +245,23 @@ def _add_default_metadata(self):
         self.metadata["command_line_arguments"] = " ".join(sys.argv)
 
 
-# Global instance of a TimerStack. This is generally all that we need for profiling, but you can potentially
-# create multiple instances and pass them to the contextmanager
-_global_timer_stack = TimerStack()
+# Maintain a separate "global" timer per thread, so that they don't accidentally conflict with each other.
+_thread_timer_stacks: Dict[int, TimerStack] = {}
+
+
+def _get_thread_timer() -> TimerStack:
+    ident = threading.get_ident()
+    if ident not in _thread_timer_stacks:
+        timer_stack = TimerStack()
+        _thread_timer_stacks[ident] = timer_stack
+    return _thread_timer_stacks[ident]
+
+
+def get_timer_stack_for_thread(t: threading.Thread) -> Optional[TimerStack]:
+    if t.ident is None:
+        # Thread hasn't started, shouldn't ever happen
+        return None
+    return _thread_timer_stacks.get(t.ident)
```

Comment:
Could we use `_thread_timer_stacks.get(t.ident, None)` with the understanding that `_thread_timers_stacks[None]` should never work?

Reply:
In practice, it's fine. But `mypy` complained about the types - the dict key is `int` but `Thread.ident` can be None if the thread hasn't started yet.